src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs

```diff
@@ -19,46 +21,82 @@ public DefaultTagHelperResolver(bool designTime)
         public override IReadOnlyList<TagHelperDescriptor> GetTagHelpers(Compilation compilation)
         {
             var results = new List<TagHelperDescriptor>();
+            var errors = new ErrorSink();
+
+            VisitTagHelpers(compilation, results, errors);
+            VisitViewComponents(compilation, results, errors);
+
+            return results;
+        }
 
-            // If ITagHelper isn't defined, then we couldn't possibly find anything.
+        private void VisitTagHelpers(Compilation compilation, List<TagHelperDescriptor> results, ErrorSink errors)
+        {
             var @interface = compilation.GetTypeByMetadataName(TagHelperTypes.ITagHelper);
             if (@interface == null)
             {
-                return results;
+                // If ITagHelper isn't defined, then we couldn't possibly find anything.
+                return;
             }
 
             var types = new List<INamedTypeSymbol>();
-            var visitor = new Visitor(@interface, types);
+            var visitor = new TagHelperVisitor(@interface, types);
 
-            visitor.Visit(compilation.Assembly.GlobalNamespace);
-
-            foreach (var reference in compilation.References)
-            {
-                if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
-                {
-                    visitor.Visit(assembly.GlobalNamespace);
-                }
-            }
+            VisitCompilation(visitor, compilation);
 
-            var errors = new ErrorSink();
             var factory = new DefaultTagHelperDescriptorFactory(compilation, DesignTime);
 
             foreach (var type in types)
             {
                 var descriptors = factory.CreateDescriptors(type, errors);
                 results.AddRange(descriptors);
             }
+        }
 
-            return results;
+        private static void VisitViewComponents(Compilation compilation, List<TagHelperDescriptor> results, ErrorSink errors)
+        {
+            var @interface = compilation.GetTypeByMetadataName(ViewComponentTypes.ViewComponentAttribute);
```

Comment:
> should we no-op if this type isn't defined?

If there's no attribute the classes could be defined in a class library using the ViewComponent suffix couldn't they?



> Have you given any thought to how we'll determine if the user is on MVC >= 1.1

Great question. Hmm, I could look for public types added in 1.1 to check for its existence. What do you think?